File: CP164/sidh9846_data_structures/src/Queue_linked.py
"""
-------------------------------------------------------
Linked version of the Queue ADT.
-------------------------------------------------------
Author:  David Brown
ID:      999999999
Email:   dbrown@wlu.ca
Term:    Winter 2020
__updated__ = "2023-03-13"
-------------------------------------------------------
"""
from copy import deepcopy


class _Queue_Node:

    def __init__(self, value, next_):
        """
        -------------------------------------------------------
        Initializes a queue node that contains a copy of value
        and a link to the next node in the queue.
        Use: node = _Queue_Node(value, _next)
        -------------------------------------------------------
        Parameters:
            value - value for node (?)
            next_ - another Queue node (_Queue_Node)
        Returns:
            a new _Queue_Node object (_Queue_Node)
        -------------------------------------------------------
        """
        self._value = deepcopy(value)
        self._next = next_


class Queue:

    def __init__(self):
        """
        -------------------------------------------------------
        Initializes an empty queue. Data is stored in a Python list.
        Use: queue = Queue()
        -------------------------------------------------------
        Returns:
            a new Queue object (Queue)
        -------------------------------------------------------
        """
        self._front = None
        self._rear = None
        self._count = 0

    def is_empty(self):
        """
        -------------------------------------------------------
        Determines if the queue is empty.
        Use: b = queue.is_empty()
        -------------------------------------------------------
        Returns:
            True if queue is empty, False otherwise.
        -------------------------------------------------------
        """
        # your code here

        is_empty = False

        if self._count == 0:
            is_empty = True

        return is_empty

    def is_full(self):
        """
        -------------------------------------------------------
        Determines if the queue is full.
        Use: b = queue.is_full()
        -------------------------------------------------------
        Returns:
            True if queue is full, False otherwise.
        -------------------------------------------------------
        """
        # your code here
        return False

    def __len__(self):
        """
        -------------------------------------------------------
        Returns the length of the queue.
        Use: n = len(queue)
        -------------------------------------------------------
        Returns:
            the number of values in queue.
        -------------------------------------------------------
        """
        # your code here

        length = self._count

        return length

    def insert(self, value):
        """
        -------------------------------------------------------
        Inserts a copy of value into the queue.
        Use: queue.insert(value)
        -------------------------------------------------------
        Parameters:
            value - a data element (?)
        Returns:
            a copy of value is added to the rear of queue.
        -------------------------------------------------------
        """
        # your code here

        node = _Queue_Node(deepcopy(value), None)

        if self._count == 0:
            self._front = node
            self._rear = self._front
        else:
            self._rear._next = node
            self._rear = node

        self._count += 1

        return

    def remove(self):
        """
        -------------------------------------------------------
        Removes and returns value from the queue.
        Use: value = queue.remove()
        -------------------------------------------------------
        Returns:
            value - the value at the front of the queue - the value is
            removed from queue (?)
        -------------------------------------------------------        
        """
        assert self._front is not None, "Cannot remove from an empty queue"

        # your code here

        value = self._front._value

        self._front = self._front._next

        if self._front is None:
            self._rear = None

        self._count -= 1

        return value

    def peek(self):
        """
        -------------------------------------------------------
        Peeks at the front of queue.
        Use: value = queue.peek()
        -------------------------------------------------------
        Returns:
            value - a copy of the value at the front of queue -
            the value is not removed from queue (?)
        -------------------------------------------------------
        """
        assert self._front is not None, "Cannot peek at an empty queue"

        # your code here

        value = deepcopy(self._front._value)

        return value

    def _move_front_to_rear(self, source):
        """
        -------------------------------------------------------
        Moves the front node from the source queue to the rear of the target queue.
        The target queue contains the old front node of the source queue.
        The source queue front is updated.
        Use: target._move_front_to_rear(source)
        -------------------------------------------------------
        Parameters:
            source - a linked queue (Queue)
        Returns:
            None
        -------------------------------------------------------
        """
        assert source._front is not None, "Cannot move the front of an empty queue"

        # your code here

        temp = source._front

        source._front = source._front._next

        temp._next = None

        if self._count == 0:
            self._front = temp
            self._rear = self._front
        else:
            self._rear._next = temp
            self._rear = temp

        self._count += 1
        source._count -= 1

        if source._count == 0:
            source._rear = None

        return

    def _append_queue(self, source):
        """
        -------------------------------------------------------
        Appends the entire source queue to the rear of the target queue.
        The source queue becomes empty.
        Use: target._append_queue(source)
        -------------------------------------------------------
        Parameters:
            source - an linked-based queue (Queue)
        Returns:
            None
        -------------------------------------------------------
        """
        assert source._front is not None, "Cannot append an empty queue"

        # your code here

        if self._count == 0:
            self._front = source._front
            self._rear = source._rear
            self._count = source._count
        else:
            self._count += source._count
            self._rear._next = source._front
            self._rear = source._rear

        source._front = None
        source._rear = None
        source._count = 0

        return

    def combine(self, source1, source2):
        """
        -------------------------------------------------------
        Combines two source queues into the current target queue. 
        When finished, the contents of source1 and source2 are interlaced 
        into target and source1 and source2 are empty.
        (iterative algorithm)
        Use: target.combine(source1, source2)
        -------------------------------------------------------
        Parameters:
            source1 - an linked queue (Queue)
            source2 - an linked queue (Queue)
        Returns:
            None
        -------------------------------------------------------
        """
        # your code here

        while source1._front is not None and source2._front is not None:
            self._move_front_to_rear(source1)
            self._move_front_to_rear(source2)

        while source1._front is not None:
            self._move_front_to_rear(source1)

        while source2._front is not None:
            self._move_front_to_rear(source2)

        return

    def split_alt(self):
        """
        -------------------------------------------------------
        Splits the source queue into separate target queues with values 
        alternating into the targets. At finish source queue is empty.
        (iterative algorithm)
        Use: target1, target2 = source.split()
        -------------------------------------------------------
        Returns:
            target1 - contains alternating values from source (Queue)
            target2 - contains other alternating values from source (Queue)
        -------------------------------------------------------
        """
        # your code here

        target1 = Queue()
        target2 = Queue()

        current = self._front

        while current is not None:
            item = self.remove()
            target1.insert(item)
            current = current._next
            if current is not None:
                item = self.remove()
                target2.insert(item)
                current = current._next

        return target1, target2

    def __eq__(self, target):
        """
        ---------------------------------------------------------
        Determines whether two Queues are equal.
        Values in self and target are compared and if all values are equal
        and in the same order, returns True, otherwise returns False.
        Use: equals = source == target
        ---------------
        Parameters:
            target - a queue (Queue)
        Returns:
            equals - True if source contains the same values
                as target in the same order, otherwise False. (boolean)
        -------------------------------------------------------
        """
        # your code here

        equals = True

        current_self = self._front
        current_target = target._front

        if self._count != target._count:
            equals = False

        while equals and current_self is not None and current_target is not None:
            if current_self._value != current_target._value:
                equals = False
            else:
                current_self = current_self._next
                current_target = current_target._next

        return equals

    def __iter__(self):
        """
        USE FOR TESTING ONLY
        -------------------------------------------------------
        Generates a Python iterator. Iterates through the queue
        from front to rear.
        Use: for v in q:
        -------------------------------------------------------
        Returns:
            value - the next value in the queue (?)
        -------------------------------------------------------
        """
        current = self._front

        while current is not None:
            yield current._value
            current = current._next

File: CP264/sidh9846_l06_BAD/src/queue_linked.c
/**
 * -------------------------------------
 * @file  queue_linked.c
 * Linked Queue Source Code File
 * -------------------------------------
 * @author David Brown, 123456789, dbrown@wlu.ca
 *
 * @version 2023-10-08
 *
 * -------------------------------------
 */
// Includes
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "queue_linked.h"

// Functions

// Initializes a queue.
void queue_initialize(queue_linked_struct **queue, data_type_copy copy,
		data_type_string to_string, data_type_destroy destroy) {

	// your code here

	*queue = malloc(sizeof **queue);
	assert((*queue) != NULL);
	(*queue)->front = NULL;
	(*queue)->rear = NULL;
	(*queue)->copy = copy;
	(*queue)->to_string = to_string;
	(*queue)->destroy = destroy;
	return;

}

// Destroys a queue.
void queue_destroy(queue_linked_struct **source) {

	// your code here

	// Free the linked data
	while ((*source)->front != NULL) {
		queue_node *temp = (*source)->front;
		(*source)->destroy(&(temp->item));
		(*source)->front = (*source)->front->next;
		free(temp);
		temp = NULL;
	}
	// Free the queue header
	free(*source);
	*source = NULL;
	return;

}

// Determines if a queue is empty.
bool queue_empty(const queue_linked_struct *source) {

	// your code here

	return (source->front == NULL);

}

// Determines if a queue is full.
bool queue_full(const queue_linked_struct *source) {
	return false;
}

// Returns number of items in queue.
int queue_count(const queue_linked_struct *source) {

	// your code here

	return source->count;

}

// Inserts a copy of an item into a queue.
bool queue_insert(queue_linked_struct *source, const data_type *item) {

	// your code here

	bool inserted = false;
	queue_node *node = malloc(sizeof *node);

	if (node != NULL) {
		node->item = malloc(sizeof *node->item);

		if (node->item != NULL) {
			source->copy(node->item, item);
			source->rear->next = node;
			node->next = NULL;
			source->rear = node;
			inserted = true;
		}
	}
	return inserted;

}

// Returns a copy of the item at the front of a queue, queue is unchanged.
bool queue_peek(const queue_linked_struct *source, data_type *item) {

	// your code here

	bool peeked = false;

	if (source->front != NULL) {
		source->copy(item, source->front->item);
		peeked = true;
	}
	return peeked;

}

// Removes and returns the item at the front of a queue.
bool queue_remove(queue_linked_struct *source, data_type *item) {

	// your code here

	bool removed = false;

	if (source->front != NULL) {
		source->copy(item, source->front->item);
		queue_node *temp = source->front;
		source->front = source->front->next;
		free(temp);
		removed = true;
	}
	return removed;

}

// Prints the items in a queue from front to rear.
void queue_print(const queue_linked_struct *source) {
	char string[DATA_STRING_SIZE];
	queue_node *current = source->front;

	while (current != NULL) {
		printf("%s\n",
				source->to_string(string, DATA_STRING_SIZE, current->item));
		current = current->next;
	}
	return;
}

static void queue_move_front(queue_linked_struct *target,
		queue_linked_struct *source) {
	assert(source->front != NULL);

	queue_node *temp = source->front;
	source->front = source->front->next;
	temp->next = target->front;
	target->front = temp;
	return;
}

// Combines contents of two source queues into a target queue. Items alternate.
void queue_combine(queue_linked_struct *target, queue_linked_struct *source1,
		queue_linked_struct *source2) {

	// your code here

	while (source1->front != NULL && source2->front != NULL) {
		queue_move_front(target, source1);
		queue_move_front(target, source2);
	}
	while (source1->front != NULL) {
		queue_move_front(target, source1);
	}
	while (source2->front != NULL) {
		queue_move_front(target, source2);
	}
	return;

}

// Splits a source queue alternately into two target queues.  Items alternate.
void queue_split_alt(queue_linked_struct *target1, queue_linked_struct *target2,
		queue_linked_struct *source) {

	// your code here

	bool left = true;

	while (source->front != NULL) {

		if (left) {
			queue_move_front(target1, source);
		} else {
			queue_move_front(target2, source);
		}
		left = !left;
	}
	return;

}
